ndless-sdk/libndls/config.c

```diff
@@ -102,73 +141,146 @@ void cfg_open_file(const char *filepath) {
+				strlcpy(entry->key, kv, 15);
+				size_t val_len = strlen(value) + 1;
+				entry->value_sz = pow2_roundup(val_len);
+				entry->value = malloc(entry->value_sz);
+				strlcpy(entry->value, value, entry->value_sz);
+				// printf("entry %s=%s\n", entry->key, entry->value);
 				if (kv_index >= max_kv_num) break; // too many key-value pairs
 			}
 		}
 		if (!end_of_pair) break; // end of file
 		str = end_of_pair; // next key value
 		if (str >= end_of_file) break; // end of file
 		if (has_comment) {
-			str = strpbrk(str, "\n");  // skip it
+			char *ns = strpbrk(str, "\n");
+			if(value) {
+				if(!ns) ns = end_of_file;
+				char *cmt = cfg_comments[kv_index - 1] = malloc(ns - str + 1);
+				memcpy(cmt, str, ns - str);
+				cmt[ns - str] = '\0';
+			}
+			str = ns;  // skip it
 			if (!str) break; // end of file
 			str++;
 		}
 	}
-	kv_num = kv_index;
+	loop_end:
+	cmt_num = kv_num = kv_index;
+	strlcpy(open_file, filepath, FILENAME_MAX);
+	free(file_content);
 }
 
+static char cfg_path[FILENAME_MAX] = {0};
+
 static int cfg_locate_cfg_file(char *dst_path, size_t dst_path_size) {
-	return locate("ndless.cfg.tns", dst_path, dst_path_size);
+	if(cfg_path[0] == 0) snprintf(cfg_path, FILENAME_MAX, "%s%s", get_documents_dir(), "ndless/ndless.cfg.tns");
+	if(access(cfg_path, F_OK) == -1) {
+		int l = locate("ndless.cfg.tns", cfg_path, FILENAME_MAX);
+		if(l != 0) return l;
+	}
+	size_t actual_path_len = strlen(cfg_path);
+	if(actual_path_len+1 > dst_path_size) return 1;
+	memcpy(dst_path, cfg_path, actual_path_len);
+	dst_path[actual_path_len] = 0;
+	return 0;
 }
 
 void cfg_open(void) {
-	char path[300];
+	char path[FILENAME_MAX];
 	if (cfg_locate_cfg_file(path, sizeof(path)))
 		return;
 	cfg_open_file(path);
 }
 
-// Returns the value associated to the key. NULL if not found.
-char *cfg_get(const char *key) {
+struct cfg_entry *cfg_get_entry(const char *key) {
 	unsigned i;
-	if (!file_content) return NULL;
+	if (!open_file[0]) return NULL;
 	for (i = 0; i < kv_num; i++) {
-		if (!strcmp(key, kv_offsets[i][0] + file_content))
-			return kv_offsets[i][1] + file_content;
+		struct cfg_entry *e = &cfg_entries[i];
+		//printf("%s, %s\n", key, e->key);
+		if (!strcmp(key, e->key)) {
+			return e;
+		}
 	}
 	return NULL;
 }
 
-// Only for tests. cfg_register_filext() should be used.
-void cfg_register_fileext_file(const char *filepath, const char *ext, const char *prgm) {
-	char key[15] = "ext.";
-	cfg_open_file(filepath);
-	strncat(key, ext, 15 - 4 - 1);
-	if (cfg_get(key)) {
-		cfg_close();
-		return;
+// Returns the value associated to the key. NULL if not found.
+const char *cfg_get(const char *key) {
+	struct cfg_entry *e = cfg_get_entry(key);
+	return e == NULL ? NULL : e->value; 
+}
+
+static int strict_strcmp(const char *a, const char *b) {
+	size_t l1 = strlen(a);
+	size_t l2 = strlen(b);
+	return memcmp(a, b, (l1 < l2 ? l1 : l2)+1);
+}
+
+void cfg_put_entry(struct cfg_entry *entr, const char *val) {
```

Comment:
This API doesn't make sense - struct cfg_entry already has a value.

Reply:
The point of this is to let config.c only write out the config if a change has actually happened. Maybe it would be better to make `struct cfg_entry` opaque outside of config.c. `cfg_put_entry` isn't a great name, I guess. Maybe cfg_set_value?

Reply:
Yes, cfg_entry sholud be opaque.